File: Assignments in CPP/21-5/hash.cpp
/*
  Patrick Morrissey pmorriss morrisseypd369@gmail.com
  hash.cpp - Assignment 05 (Hash Table)
*/

#include <iostream>
#include <iomanip>
#include <fstream>
#include <vector>
#include <string>
#include <utility>
#include "hash.h"

  hashTable::hashTable(int size){
    m = size;
    A = new std::vector<std::pair<int, std::string>>[m];
  }

  hashTable::~hashTable() {
    delete[] A;
  }

  void hashTable::hash_insert(int key, std::string value) {
    int hash_key = hash_generate(key);
    std::pair<int, std::string> data = std::make_pair(key, value);
    A[hash_key].push_back(data);
  }

  std::string hashTable::hash_delete(int key) {
    int hash_key = hash_generate(key);
    std::string value;
    std::pair <int, std::string> data;
    int index = find(key, hash_key);   
    if(index != -1) {
      data = A[hash_key][index];
      value = data.second;
      int temp = A[hash_key].size()-1;
      A[hash_key][temp].swap(A[hash_key][index]);
      A[hash_key].pop_back();
      return value;
    }
    return "";
  }

  void hashTable::hash_load(std::string filename) {
    std::ifstream fin(filename);
    int key;
    std::string input;
    while(std::getline(fin, input)) {
      std::string key_temp;
      std::string value;
      for (int i = 0; i < input.size(); ++i) {
       if(i < input.find_first_of(' ')) {
          key_temp += input[i];
        } else if (i > input.find_first_of(' ')) {
          value += input[i];
        }
      }
      key = std::stoi(key_temp);
      hash_insert(key, value);
    }
    fin.close();
  }

  void hashTable::hash_save(std::string filename) {
    std::ofstream fout(filename);
    if(fout.is_open()){
      for(int i = 0; i < m; ++i){
        for(int index = 0; index < A[i].size(); ++index) {
          std::pair<int, std::string> data = A[i][index];
          fout << std::setfill('0') << std::setw(9) << data.first 
          << " " << data.second << std::endl;
        }
      }
    } else {
      std::cout << "Error opening file! Please try again" << std::endl;
    }
    fout.close();
  }

  void hashTable::hash_clear() {
    for(int i = 0; i < m; ++i) {
      A[i].clear();
    }
  }

  inline int hashTable::find(int key, int hash_key) {
    if(A[hash_key].size() > 0) {
      for(int i = 0; i < A[hash_key].size(); ++i) {
        std::pair <int, std::string> data = A[hash_key][i];
        if(data.first == key) {
          return i;
        }
      }
    }
    return -1;
  }

File: Assignments in Java/21-5/hash.java
import java.util.Vector;
import java.util.Scanner;
import java.io.File;
import java.io.FileWriter;
import java.io.FileNotFoundException;
import java.io.IOException;

public class hash {
  Integer m;
  final double c = 0.618034;
  Vector<Vector<pair>> A = new Vector<Vector<pair>>();
  

  public hash(Integer size) {
    m = size;
    for (int i = 0; i < m; i++) { // resizes the hash table
      A.add(new Vector<pair>());
    }
  }

  public void hash_insert(Integer key, String value) {
    Integer hash_key = hash_generate(key);
    A.get(hash_key).add(new pair(key, value));
  }

  public String hash_delete(Integer key) {
    Integer hash_key = hash_generate(key);
    Integer index = find(key, hash_key);
    if(index != -1) {
      String value = A.get(hash_key).get(index).second();
      System.out.println("found index at: " + index);
      A.get(hash_key).removeElementAt(index);
      return value;
    }
    return "";
  }

  public String hash_search(Integer key) {
    Integer hash_key = hash_generate(key);
    Integer index = find(key, hash_key);
    if(index != -1) {
      return A.get(hash_key).get(index).second();
    }
    return "";
  }

  public void hash_load(String filename) {
    File fin = new File(filename);
    try {
      Scanner scanner = new Scanner(fin);
      while(scanner.hasNextLine()) {
        Integer key = scanner.nextInt();
        String value = scanner.nextLine().trim();
        hash_insert(key, value);
      }
      scanner.close();
    } catch (FileNotFoundException e) {
      System.out.println("File not found: " + filename);
    }
  }

  public void hash_save(String filename) {
    try {
      FileWriter fout = new FileWriter(filename);
      for(Integer i = 0; i < m; ++i){
        for(Integer index = 0; index < A.get(i).size(); ++i) {
          fout.write(String.format("%09d", A.get(i).get(index).first()) + " " + A.get(i).get(index).second() + "\n");
        }
      }
      fout.close();
    } catch (IOException e) {
      System.out.println("An error occurred!");
    }

  }

  public void hash_clear() {
    for(int i = 0; i < m; ++i) {
      A.get(i).clear();
    }
  }

  private Integer find(Integer key, Integer hash_key) {
    if(A.get(hash_key).size() > 0) {
      for(Integer i = 0; i < A.get(hash_key).size(); ++i) {
        if(A.get(hash_key).get(i).first() == key) {
          return i;
        }
      }
    }
    return -1;
  }

  private Integer hash_generate(Integer key) {
    return (int)(m*(key*c - (int)(key*c)));
  }
}
